### PR TITLE
Add lint `equatable_if_let`

### PR DESCRIPTION
**Pull request by HKalbasi**
This is my attempt for #1716. There is a major false positive, which is people may implement `PartialEq` in a different way. It is unactionable at the moment so I put it into `nursery`.

There is a trait `StructuralPartialEq` for solving this problem which is promising but it has several problems currently:
* Integers and tuples doesn't implement it.
* Some types wrongly implement it, like `Option<T>` when `T` doesn't implement it.

I consider them bugs and against the propose of `StructuralPartialEq`. When they become fixed, this lint can become a useful lint with a single line change.

changelog: New lint: [`equatable_if_let`]

### CODE REVIEW

**Review by Manishearth — APPROVED** [review]

**Comment by camsteffen** [comment]
name idea: equatable_if_let

File: clippy_lints/src/pattern_equality.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    /// ```
+    pub PATTERN_EQUALITY,
+    nursery,
+    "using pattern matching instead of equality"
+}
+
+declare_lint_pass!(PatternEquality => [PATTERN_EQUALITY]);
+
+/// detects if pattern matches just one thing
+fn unary_pattern(pat: &Pat<'_>) -> bool {
```
[the remark this answers is not in the document]
Comment (camsteffen) [reply clippy_lints/src/pattern_equality.rs:43]: I think this is pretty specific to this lint - "patterns that could possibly be used as an rvalue"

[the remark this answers is not in the document]
Comment (camsteffen) [reply clippy_lints/src/pattern_equality.rs:43]: ...well I suppose that can be a util. We don't really have a module for it other than top-level clippy_utils.

**Comment by HKalbasi** [comment]
Originally extending it to `matches!` was in my mind (comment on this?) and single match is mentioned in the original issue, so maybe `equatable_pattern_matching` or `equatable_pattern`?

File: clippy_lints/src/pattern_equality.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (HKalbasi) [reply clippy_lints/src/pattern_equality.rs:43]: I'm fan of putting functions in utils, but I can't imagine any other use case for this function.

**Comment by camsteffen** [comment]
Linting `matches!` seems more picky to me so I wouldn't. The single match case is already linted to be changed to `if let`, so I consider `if let` to be the "base case" of this lint.

**Comment by HKalbasi** [comment]
Make sense :+1:

Is there any tool for renaming lint? They are everywhere in variety of cases.

**Comment by camsteffen** [comment]
No but that's a good idea for clippy dev!

File: clippy_lints/src/pattern_equality.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Manishearth) [reply clippy_lints/src/pattern_equality.rs:43]: Okay, that works for me

**Comment by HKalbasi** [comment]
I'm still renaming the lint :)

**Comment by camsteffen** [comment]
The changelog diff is massive

**Comment by HKalbasi** [comment]
Ah, my formatter seems don't like it. Now it is fixed.

**Pull request merged**